streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/HighAvailabilityTaskAssignor.java

```diff
@@ -0,0 +1,544 @@
+
+    private final AssignmentConfigs configs;
+
+    private final SortedMap<TaskId, SortedSet<RankedClient<ID>>> statefulTasksToRankedCandidates;
+
+    public HighAvailabilityTaskAssignor(final Map<ID, ClientState> clientStates,
+                                 final Set<TaskId> allTasks,
+                                 final Set<TaskId> statefulTasks,
+                                 final AssignmentConfigs configs) {
+        this.configs = configs;
+        this.clientStates = clientStates;
+        this.allTasks = allTasks;
+        this.statefulTasks = new TreeSet<>(statefulTasks);
+
+        statelessTasks = new TreeSet<>(allTasks);
+        statelessTasks.removeAll(statefulTasks);
+
+        sortedClients = new TreeSet<>();
+        clientsToNumberOfThreads = new HashMap<>();
+        clientStates.forEach((client, state) -> {
+            sortedClients.add(client);
+            clientsToNumberOfThreads.put(client, state.capacity());
+        });
+
+        statefulTasksToRankedCandidates =
+            buildClientRankingsByTask(statefulTasks, clientStates, configs.acceptableRecoveryLag);
+    }
+
+    @Override
+    public boolean assign() {
+        final Map<ID, List<TaskId>> warmupTaskAssignment = initializeEmptyTaskAssignmentMap();
+        final Map<ID, List<TaskId>> standbyTaskAssignment = initializeEmptyTaskAssignmentMap();
+        final Map<ID, List<TaskId>> statelessActiveTaskAssignment = initializeEmptyTaskAssignmentMap();
+
+        // ---------------- Stateful Active Tasks ---------------- //
+
+        final Map<ID, List<TaskId>> statefulActiveTaskAssignment =
+            new DefaultStateConstrainedBalancedAssignor<ID>().assign(
+                statefulTasksToRankedCandidates,
+                configs.balanceFactor,
+                sortedClients,
+                clientsToNumberOfThreads
+            );
+
+        // ---------------- Warmup Replica Tasks ---------------- //
+
+        final Map<ID, List<TaskId>> balancedStatefulActiveTaskAssignment =
+                new DefaultBalancedAssignor<ID>().assign(
+                    sortedClients,
+                    statefulTasks,
+                    clientsToNumberOfThreads,
+                    configs.balanceFactor);
+
+        final Queue<Movement<ID>> movements =
+            getMovements(statefulActiveTaskAssignment, balancedStatefulActiveTaskAssignment);
+        for (int numWarmupReplicas = 0; numWarmupReplicas < configs.maxWarmupReplicas; ++numWarmupReplicas) {
+            final Movement<ID> movement = movements.poll();
+            if (movement == null) {
+                break;
+            }
+            warmupTaskAssignment.get(movement.destination).add(movement.task);
+        }
+
+        // ---------------- Standby Replica Tasks ---------------- //
+
+        final List<Map<ID, List<TaskId>>> allStatefulTaskAssignments = asList(
+            statefulActiveTaskAssignment,
+            warmupTaskAssignment,
+            standbyTaskAssignment
```

Comment:
I think I might be missing something... The `standbyTaskAssignment` is empty here, and also on Line 130.
Why do we need to add it to these collections?

Reply:
It's empty when we create the queue, but starts to fill up as we use the queue. The queue needs a reference to all stateful assignment maps so it can verify whether a task can be assigned to that client. The queue also needs a reference to the two standby-type assignment maps (as in Line 130) so it choose the next least loaded client when polled.

Reply:
I was worried this part of the code wasn't all that clear, this seems like an indication that may be the case...I'm open to suggestions, or to pushing that to followup work if that makes more sense

Reply:
sounds good. Let's put it off for now.